### DISCUSSION TITLE
Luau Module

### DISCUSSION THREAD

**Discussion by DasLixou** [mlua-rs/mlua#286] [Q&A]
How can I export a function / make a module with luau?
I get `compile_error!("Luau does not support module mode");`
Is there any possible way to make luau call rust code?

**Comment by khvzak** [comment]
Luau itself does not support loading binary shared libraries.
In theory it should be possible, but I need to add support to mlua calling [dlopen](https://man7.org/linux/man-pages/man3/dlopen.3.html) to load module symbols.
I was considering adding this functionality, but not sure will it be in 0.9 or later.

**Reply by DasLixou** [reply]
So there's currently no possible way of doing modules? I'm new to lua so this idea might be stupid but when trying to understand the module thing.. couldn't I just make a table and instead of making it a module I just smash it in a global?

**Reply by filiptibell** [reply]
Adding module support for Luau would be very useful, for my project [Lune](https://github.com/filiptibell/lune) which is a standalone runtime based on Luau, I would like to let people create native speed modules, `require` them as normal, but it's currently not possible. This leads to me having to integrate any functionality that needs native code into the full runtime but I would like the runtime to remain pretty small.

Also, a related question - would it be possible to make `mlua` modules portable? So that a user can distribute a single file / package and any os/arch that is supported by mlua can run it?

I think support for this would be amazing and now that 0.9 is out I hope you work on it 👍

**Reply by khvzak** [reply]
I work on native modules for Luau, should be in the next version.

>  would it be possible to make mlua modules portable?

I don't think this is a question for mlua, rather for core Rust/LLVM :)

**Reply by khvzak** [reply]
This is supported now since v0.9.2
